db.py

```diff
@@ -1,82 +1,114 @@
+        count = self.getCount(url)
         if count == 0:
             cursor.execute('INSERT INTO url(url, count) VALUES(?, ?)', (url, 0))
         # Add 1 to the url count
-        cursor.execute('UPDATE url SET count = count + 1 WHERE url=?', (url, ))
-        connection.commit()
+        cursor.execute('UPDATE url SET count = count + 1 WHERE url=?', (url,))
+        self.connection.commit()
 
-    def getTopSites(self, connection, amount=10):
+    def getTopSites(self, amount=10):
         """ Get the top domains using this tool by hits. Ignore specified domains """
-        # Select all urls and counts
-        cursor = connection.cursor()
-        cursor.execute('select url, count from url')
-        urls_and_counts = cursor.fetchall()
-
-        # Get total hits per domain
-        site_counts = defaultdict(int)
-        for row in urls_and_counts:
-            if row[0] == b'':
-                continue
-            # Get the domain - part before the first '/'
-            domain = row[0].split('/')[0]
-            # Check if domain is on the ignore list
-            on_ignore = False
-            for regex in config.TOP_SITES_IGNORE_DOMAIN_RE_MATCH:
-                if re.match(regex, domain) is not None:
-                    on_ignore = True
-                    break
-            if on_ignore:
-                continue
-            # Add hit counts to the domain
-            site_counts[domain] += row[1]
+        return self.getTop(amount, 'domains')
 
-        # Sort the domains by hits
-        sorted_sites = sorted(site_counts, key=lambda x: site_counts[x], reverse=True)
+    def getTopUrls(self, amount=10):
+        """ Get the top urls using this tool by hits. Ignore specified domains """
+        return self.getTop(amount, 'urls')
 
-        # Return sorted domains and their values, this allows for lower Python version support
+    def getTop(self, amount, what='domains'):
+        query = self._top_urls_query() if what == 'urls' else self._top_domains_query()
+        # Select all entities and counts
+        cursor = self.connection.cursor()
+        cursor.execute(
+            query,
+            [*[str(r) for r in config.TOP_SITES_IGNORE_DOMAIN_RE_MATCH], amount]
+        )
+        result = cursor.fetchall()
+
+        entities, values = [], {}
+        for row in result:
+            url, domain, count = row
+            if url == b'':
+                continue
+            entities.append(url)
+            values[url] = count
+
+        # Return sorted entities and their values, this allows for lower Python version support
         return {
-            'domains': sorted_sites[:amount],
-            'values': {site: site_counts[site] for site in site_counts}
+            what: entities,
+            'values': values
         }
+
+    def _top_domains_query(self):
+        sep = '\nAND '
+        return f"""
+            SELECT url, substr(url, 0, instr((url || '/'), '/')) as domain, SUM(count) as domain_sum
```

Comment:
We are grouping by domain and then still returning a URL column - does this hold any useful data? Maybe you could just return two columns; the entity and count.

Reply:
It was only for convenience to not require a case distinction [here](https://github.com/brentvollebregt/hit-counter/pull/9/commits/e7ed534c5d707868b133341b9b9cf53e975ddf65#diff-e89093fa2820829fa461a51870e78e5dR80). But as it would probably cost performance, I changed it now.